### PR TITLE
Keep workers records when they gone offline

### PR DESCRIPTION
**Pull request by BrnoPCmaniak**
closes #2430
https://pulp.plan.io/issues/2430

### CODE REVIEW

**Comment by dralley** [comment]
I'll get to this within the next few days, hopefully by the weekend

File: platform/pulp/app/models/task.py [hunk 1]
```diff
@@ -70,20 +70,23 @@ class Worker(Model):
...
         """Save a worker heartbeat
 
         Update the last_heartbeat field to now and save it.
 
-        Warning:
+        Warnings:
 
-            Only the last_heartbeat field will be saved. No other changes will be saved.
+            1) Only the last_heartbeat field will be saved. No other changes will be saved.
+            2) If you will try to save via this method new model it will lead to crash.
```
Comment (bmbouter) [inline platform/pulp/app/models/task.py]: I think this line 2) should be converted into a `Raises` statement like this:  https://github.com/BrnoPCmaniak/pulp/blob/bd1f983ab32ba94d0b44e3583baa9ee0a18ee468/platform/pulp/app/models/task.py#L54

File: platform/pulp/app/serializers/task.py [hunk 2]
```diff
@@ -94,6 +94,11 @@ class WorkerSerializer(ModelSerializer):
         read_only=True
     )
 
+    online = serializers.BooleanField(
+        help_text='Whether is the worker online or not. Default is True.',
```
Comment (bmbouter) [inline platform/pulp/app/serializers/task.py]: s/Whether is the worker online or not. Default is/Whether the worker is online or not. Defaults to/

File: platform/pulp/tasking/celery_app.py [hunk 3]
```diff
@@ -124,11 +124,12 @@ def initialize_worker(sender, instance, **kwargs):
     We clean up old state in case this worker was previously running, but died unexpectedly.
     In such cases, any Pulp tasks that were running or waiting on this worker will show incorrect
     state. Any reserved_resource reservations associated with the previous worker will also be
-    removed along with the worker entry in the database itself. The working directory specified in
-    /etc/pulp/server.conf (/var/cache/pulp/<worker_name>) by default is removed and recreated. This
-    is called early in the worker start process, and later when it's fully online, pulp_celerybeat
-    will discover the worker as usual to allow new work to arrive at this worker. If there is no
-    previous work to cleanup, this method still runs, but has no effect on the database.
+    removed along with the online status of worker in the database itself. The working directory
```
Comment (bmbouter) [inline platform/pulp/tasking/celery_app.py]: s/the online status of worker/the setting of online=False on the worker record/

File: platform/pulp/tasking/services/worker_watcher.py [hunk 4]
```diff
@@ -42,6 +42,11 @@ def handle_worker_heartbeat(worker_name):
     if created:
         msg = _("New worker '{name}' discovered").format(name=worker_name)
         _logger.info(msg)
+    elif existing_worker.online is False:
+        msg = _("Worker '{name}' is now back online.").format(name=worker_name)
```
Comment (bmbouter) [inline platform/pulp/tasking/services/worker_watcher.py]: s/now //

File: platform/pulp/app/models/task.py [hunk 5]
```diff
@@ -70,20 +70,26 @@ class Worker(Model):
...
         Update the last_heartbeat field to now and save it.
 
-        Warning:
+        Warnings:
 
-            Only the last_heartbeat field will be saved. No other changes will be saved.
+            1) Only the last_heartbeat field will be saved. No other changes will be saved.
+
+        Raises:
+            ValueError("Cannot force an update in save() with no primary key."): You cannot create
```
Comment (BrnoPCmaniak) [inline platform/pulp/app/models/task.py]: @bmbouter Like this?

Comment (bmbouter) [inline platform/pulp/app/models/task.py]: Like this:

```
Raises:
    ValueError: When the model instance has never been saved before. This method can only update an existing database record.
```

**Pull request merged**